src/porepy/numerics/ad/_ad_utils.py

```diff
@@ -447,6 +447,51 @@ def get_solution_values(
     return value
 
 
+def shift_solution_values(
+    name: str,
+    data: dict,
+    location: Any,
+) -> None:
+    """Function to shift numerical values stored in the data dictionary.
+
+    The shift is implemented s.t. values at index ``i`` are copied to index ``i + 1``,
+    based on the number of stored values.
+
+    Note:
+        The data stored must have support for ``.copy()`` in order to avoid faulty
+        referencing (e.g., numpy arrays or sparse matrices).
```

Comment:
This is probably the safer option, and certainly consistent with existing code, so fine with me. I'm not sure we have the right data model, using copies all over the place, but that is a different matter.